File: Blind75/trees/subtree of another tree.py
"""edge cases:
2 helper functions - sameTree and isSubTree
If both trees are null - same tree & is a sub tree
Extra node for one of the sub trees - not same tree
Is null tree a sub tree of other tree - yes
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:return True
        if not root:return False

        if self.sameTree(root, subRoot):
            return True
        
        return (self.isSubtree(root.left,subRoot)) or self.isSubtree(root.right,subRoot)

    def sameTree(self, root,subRoot):
        if not root and not subRoot:
            return True
        if root and subRoot and root.val == subRoot.val:
            return (self.sameTree(root.left, subRoot.left) and self.sameTree(root.right, subRoot.right))

File: 150/subtree_of_another_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!root && subRoot) return false;
        if (!subRoot) return true; // subRoot is null then its a sub tree of root

        // check if same tree
        if ((isSameTree(root, subRoot))){
            return true;ac
        }
        
        // recurssive check for subnodes in root
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }

private:
        bool isSameTree(TreeNode* p, TreeNode* q){
            // base cases
            // if both are null
            if (!p && !q) return true;

            // if both are not null and root values are same then check for subnode values
            if (p && q && p->val == q->val){
            // recurssive
                return (isSameTree(p->left, q->left) and isSameTree(p->right, q->right));
            }
            // if either is empty
            return false;
    }
};
